### PR TITLE
BUG: reindex with empty inputs

### LINKED ISSUE #27315: some argument combinations with reindex fails on an empty dataframe
```python
import pandas as pd
from datetime import datetime, timedelta

end = datetime.utcnow()
begin = end - timedelta(minutes=1)
data_interval = 10
date_index = pd.date_range(start=begin,
                           end=end,
                           freq='{} s'.format(data_interval))
df = pd.DataFrame([], columns=['time','a','b'])
df = df.set_index('time', drop=True)
tol = timedelta(seconds=9)
df = df.reindex(date_index, method='pad', tolerance=tol)

# IndexError: index -1 is out of bounds for axis 0 with size 0

df = pd.DataFrame([], columns=['time','a','b'])
df = df.reindex(date_index, method='nearest')

# IndexError: index -1 is out of bounds for axis 0 with size 0

```
You get an index error when a dataframe is empty using the `tolerance=` or `method='nearest'` .

This is not something that happens with other usages of reindex and can come up as a surprise when reindexing an empty window of data. I would expect it to behave the same as it does without tolerance here.

#### Expected Output
Should be same as reindex with no args in this case which returns...
```
                              a    b
2019-07-09 22:35:05.165640  NaN  NaN
2019-07-09 22:35:15.165640  NaN  NaN
2019-07-09 22:35:25.165640  NaN  NaN
2019-07-09 22:35:35.165640  NaN  NaN
2019-07-09 22:35:45.165640  NaN  NaN
2019-07-09 22:35:55.165640  NaN  NaN
2019-07-09 22:36:05.165640  NaN  NaN
```

#### Temp Solution
Simple user solution is to check length... but this is a problem that might surprise someone at a bad time like it did for us.
``` python
if(len(df) is 0):
    df = df.reindex(date_index)
else:
    df = df.reindex(date_index, method='pad', tolerance=tol)
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-54-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 5.0.1
pip: 18.0
setuptools: 40.4.1
Cython: None
numpy: 1.16.4
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: 1.1.8
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: 1.3.5
pymysql: 0.9.3
psycopg2: None
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: 0.6.1
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by ajspera**
…seless in that case anyway

- [x] closes #27315
- [ ] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
pls always add tests

**Comment by jreback** [comment]
find the reindex tests; we need to run all combinations of args for an empty

**Comment by ajspera** [comment]
Will do when I find some more time... New to the python world and this project so I need a few more moments to get familiar with the testing.

**Comment by ajspera** [comment]
@jreback, could the tests be a while loop on a dict list using **kwargs on the method or are you looking for explicit individual tests for each arg combination? The former makes sense to me for efficiency, but being relatively new to python projects I'm not sure whether that overly obfuscates failure messages. Also I noticed the entry point on reindex tests are not on the class where I discovered the issue or even in the stack trace of those errors, and the class that I changed is further down inheritance of the tested class in there. Should we write tests for higher level DataFrame.reindex or should we go for the lowest public granularity of where (assuming private methods are not test targets like other languages) the changes are at Index.reindex? I can see reasons for either decision based on skimming through the class inheritance patterns.

Sorry if these are remedial questions, just want to make my limited time focused and productive while reducing review cycle tedium for you where my knowledge gaps are. I'll make decisions once I get into it either way if there's no preference, but some foresight would save some time if there are preferences.

Also sorry for the delay, too many commitments on my plate at the moment. Next step for this is to get a list of args for the test which is simple but tedious enough to conflict my schedule. If anyone wants to assist I can probably fit resolving any test failures into my schedule easier. If nobody claims that testing task I will wrap this up as soon as I can.

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3062,6 +3062,9 @@ def _get_nearest_indexer(self, target, limit, tolerance):
         values that can be subtracted from each other (e.g., not strings or
         tuples).
         """
+        if len(self) == 0:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: use not len

**Review by jreback — CHANGES_REQUESTED** [review]
needs a test & a whatsnew note

**Comment by jreback** [comment]
closing as stale; this does fix the issue but needs tests. if you want to continue, pls merge master and ping to re-open.

**Pull request closed without merging**